README
File::ShareDir2

The intent of File::ShareDir2 is to provide a companion to
Class::Inspector and File::HomeDir, modules that take a process that is
well-known by advanced Perl developers but gets a little tricky, and
make it more available to the larger Perl community.

Quite often you want or need your Perl module (CPAN or otherwise) to
have access to a large amount of read-only data that is stored on the
file-system at run-time.

On a linux-like system, this would be in a place such as /usr/share,
however Perl runs on a wide variety of different systems, and so the use
of any one location is unreliable.

Perl provides a little-known method for doing this, but almost nobody is
aware that it exists. As a result, module authors often go through some
very strange ways to make the data available to their code.

The most common of these is to dump the data out to an enormous Perl
data structure and save it into the module itself. The result are
enormous multi-megabyte .pm files that chew up a lot of memory
needlessly.

Another method is to put the data "file" after the __DATA__ compiler tag
and limit yourself to access as a filehandle.

The problem to solve is really quite simple.

  1. Write the data files to the system at install time.
  
  2. Know where you put them at run-time.

Perl's install system creates an "auto" directory for both every
distribution and for every module file.

These are used by a couple of different auto-loading systems to store
code fragments generated at install time, and various other modules
written by the Perl "ancient masters".

But the same mechanism is available to any dist or module to store any
sort of data.

  Using Data in your Module

"File::ShareDir2" forms one half of a two part solution.

Once the files have been installed to the correct directory, you can use
"File::ShareDir2" to find your files again after the installation.

For the installation half of the solution, see Module::Install and its
"install_share" directive.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by Adam Kennedy <adamk@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Perl extension File-ShareDir2

0.001000 2015-07-12T10:40:33Z bf23cc0
 - Initial fork from File::ShareDir


t/lib/ShareDir2.pm
package t::lib::ShareDir2;

use strict;

use vars qw{$VERSION @ISA};

BEGIN {
	$VERSION = '1.01';
	@ISA     = 'File::ShareDir2';
}

1;


lib/File/ShareDir2.pm
package File::ShareDir2;

# vim: noet ts=2 sw=2
# ABSTRACT: Locate per-dist and per-module shared files

#pod =head1 SYNOPSIS
#pod
#pod   use File::ShareDir2 ':ALL';
#pod   
#pod   # Where are distribution-level shared data files kept
#pod   $dir = dist_dir('File-ShareDir2');
#pod   
#pod   # Where are module-level shared data files kept
#pod   $dir = module_dir('File::ShareDir2');
#pod   
#pod   # Find a specific file in our dist/module shared dir
#pod   $file = dist_file(  'File-ShareDir2',  'file/name.txt');
#pod   $file = module_file('File::ShareDir2', 'file/name.txt');
#pod   
#pod   # Like module_file, but search up the inheritance tree
#pod   $file = class_file( 'Foo::Bar', 'file/name.txt' );
#pod
#pod =head1 DESCRIPTION
#pod
#pod The intent of L<File::ShareDir2> is to provide a companion to
#pod L<Class::Inspector> and L<File::HomeDir>, modules that take a
#pod process that is well-known by advanced Perl developers but gets a
#pod little tricky, and make it more available to the larger Perl community.
#pod
#pod Quite often you want or need your Perl module (CPAN or otherwise)
#pod to have access to a large amount of read-only data that is stored
#pod on the file-system at run-time.
#pod
#pod On a linux-like system, this would be in a place such as /usr/share,
#pod however Perl runs on a wide variety of different systems, and so
#pod the use of any one location is unreliable.
#pod
#pod Perl provides a little-known method for doing this, but almost
#pod nobody is aware that it exists. As a result, module authors often
#pod go through some very strange ways to make the data available to
#pod their code.
#pod
#pod The most common of these is to dump the data out to an enormous
#pod Perl data structure and save it into the module itself. The
#pod result are enormous multi-megabyte .pm files that chew up a
#pod lot of memory needlessly.
#pod
#pod Another method is to put the data "file" after the __DATA__ compiler
#pod tag and limit yourself to access as a filehandle.
#pod
#pod The problem to solve is really quite simple.
#pod
#pod   1. Write the data files to the system at install time.
#pod   
#pod   2. Know where you put them at run-time.
#pod
#pod Perl's install system creates an "auto" directory for both
#pod every distribution and for every module file.
#pod
#pod These are used by a couple of different auto-loading systems
#pod to store code fragments generated at install time, and various
#pod other modules written by the Perl "ancient masters".
#pod
#pod But the same mechanism is available to any dist or module to
#pod store any sort of data.
#pod
#pod =head2 Using Data in your Module
#pod
#pod C<File::ShareDir2> forms one half of a two part solution.
#pod
#pod Once the files have been installed to the correct directory,
#pod you can use C<File::ShareDir2> to find your files again after
#pod the installation.
#pod
#pod For the installation half of the solution, see L<Module::Install>
#pod and its C<install_share> directive.
#pod
#pod =head1 FUNCTIONS
#pod
#pod C<File::ShareDir2> provides four functions for locating files and
#pod directories.
#pod
#pod For greater maintainability, none of these are exported by default
#pod and you are expected to name the ones you want at use-time, or provide
#pod the C<':ALL'> tag. All of the following are equivalent.
#pod
#pod   # Load but don't import, and then call directly
#pod   use File::ShareDir2;
#pod   $dir = File::ShareDir2::dist_dir('My-Dist');
#pod   
#pod   # Import a single function
#pod   use File::ShareDir2 'dist_dir';
#pod   dist_dir('My-Dist');
#pod   
#pod   # Import all the functions
#pod   use File::ShareDir2 ':ALL';
#pod   dist_dir('My-Dist');
#pod
#pod All of the functions will check for you that the dir/file actually
#pod exists, and that you have read permissions, or they will throw an
#pod exception.
#pod
#pod =cut

use 5.006;    # our
use strict;
use warnings;

use Carp             ();
use Config           ();
use Exporter         ();
use File::Spec       ();
use Class::Inspector ();

our $VERSION = '0.001000';
our @ISA       = 'Exporter';
our @EXPORT_OK = qw{
	dist_dir
	dist_file
	module_dir
	module_file
	class_file
};
our %EXPORT_TAGS = ( ALL => [@EXPORT_OK], );

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

require File::ShareDir2::Resolver::Perl;
require File::ShareDir2::Resolver::Perl::Legacy;
our @RESOLVERS = ( File::ShareDir2::Resolver::Perl->new(), File::ShareDir2::Resolver::Perl::Legacy->new(), );

#####################################################################
# Interface Functions

#pod =pod
#pod
#pod =head2 dist_dir
#pod
#pod   # Get a distribution's shared files directory
#pod   my $dir = dist_dir('My-Distribution');
#pod
#pod The C<dist_dir> function takes a single parameter of the name of an
#pod installed (CPAN or otherwise) distribution, and locates the shared
#pod data directory created at install time for it.
#pod
#pod Returns the directory path as a string, or dies if it cannot be
#pod located or is not readable.
#pod
#pod =cut

sub dist_dir {
	my $dist = _DIST(shift);
	for my $resolver (@RESOLVERS) {
		my $result = $resolver->dist_dir($dist);
		if ( defined $result ) {
				return $result;
		}
	}

	# Ran out of options
	Carp::croak("Failed to find share dir for dist '$dist'");
}

#pod =pod
#pod
#pod =head2 module_dir
#pod
#pod   # Get a module's shared files directory
#pod   my $dir = module_dir('My::Module');
#pod
#pod The C<module_dir> function takes a single parameter of the name of an
#pod installed (CPAN or otherwise) module, and locates the shared data
#pod directory created at install time for it.
#pod
#pod In order to find the directory, the module B<must> be loaded when
#pod calling this function.
#pod
#pod Returns the directory path as a string, or dies if it cannot be
#pod located or is not readable.
#pod
#pod =cut

sub module_dir {
	my $module = _MODULE(shift);
	for my $resolver (@RESOLVERS) {
		my $dir = $resolver->module_dir($module);
		if ( defined $dir ) {
			return $dir;
		}
	}
	Carp::croak("Failed to find share dir for module '$module'");
}

#pod =pod
#pod
#pod =head2 dist_file
#pod
#pod   # Find a file in our distribution shared dir
#pod   my $dir = dist_file('My-Distribution', 'file/name.txt');
#pod
#pod The C<dist_file> function takes two params of the distribution name
#pod and file name, locates the dist dir, and then finds the file within
#pod it, verifying that the file actually exists, and that it is readable.
#pod
#pod The filename should be a relative path in the format of your local
#pod filesystem. It will simply added to the directory using L<File::Spec>'s
#pod C<catfile> method.
#pod
#pod Returns the file path as a string, or dies if the file or the dist's
#pod directory cannot be located, or the file is not readable.
#pod
#pod =cut

sub dist_file {
	my $dist = _DIST(shift);
	my $file = _FILE(shift);

	for my $resolver (@RESOLVERS) {
		my $path = $resolver->dist_file( $dist, $file );
		if ( defined $path ) {
			return $path;
		}
	}
	Carp::croak("Failed to find shared file '$file' for dist '$dist'");
}

#pod =pod
#pod
#pod =head2 module_file
#pod
#pod   # Find a file in our module shared dir
#pod   my $dir = module_file('My::Module', 'file/name.txt');
#pod
#pod The C<module_file> function takes two params of the module name
#pod and file name. It locates the module dir, and then finds the file within
#pod it, verifying that the file actually exists, and that it is readable.
#pod
#pod In order to find the directory, the module B<must> be loaded when
#pod calling this function.
#pod
#pod The filename should be a relative path in the format of your local
#pod filesystem. It will simply added to the directory using L<File::Spec>'s
#pod C<catfile> method.
#pod
#pod Returns the file path as a string, or dies if the file or the dist's
#pod directory cannot be located, or the file is not readable.
#pod
#pod =cut

sub module_file {
	my $module = _MODULE(shift);
	my $file   = _FILE(shift);
	for my $resolver (@RESOLVERS) {
		my $path = $resolver->module_file( $module, $file );
		if ( defined $path ) {
			return $path;
		}

	}
	Carp::croak("Failed to find shared file '$file' for module '$module'");
}

#pod =pod
#pod
#pod =head2 class_file
#pod
#pod   # Find a file in our module shared dir, or in our parent class
#pod   my $dir = class_file('My::Module', 'file/name.txt');
#pod
#pod The C<module_file> function takes two params of the module name
#pod and file name. It locates the module dir, and then finds the file within
#pod it, verifying that the file actually exists, and that it is readable.
#pod
#pod In order to find the directory, the module B<must> be loaded when
#pod calling this function.
#pod
#pod The filename should be a relative path in the format of your local
#pod filesystem. It will simply added to the directory using L<File::Spec>'s
#pod C<catfile> method.
#pod
#pod If the file is NOT found for that module, C<class_file> will scan up
#pod the module's @ISA tree, looking for the file in all of the parent
#pod classes.
#pod
#pod This allows you to, in effect, "subclass" shared files.
#pod
#pod Returns the file path as a string, or dies if the file or the dist's
#pod directory cannot be located, or the file is not readable.
#pod
#pod =cut

sub class_file {
	my $module = _MODULE(shift);
	my $file   = _FILE(shift);

	# Get the super path ( not including UNIVERSAL )
	# Rather than using Class::ISA, we'll use an inlined version
	# that implements the same basic algorithm.
	my @path  = ();
	my @queue = ( $module );
	my %seen  = ( $module => 1 );
	while ( my $cl = shift @queue ) {
		push @path, $cl;
		no strict 'refs';
		unshift @queue, grep { ! $seen{$_}++ }
			map { s/^::/main::/; s/\'/::/g; $_ }
			( @{"${cl}::ISA"} );
	}

	# Search up the path
	foreach my $class ( @path ) {
		local $@;
		my $dir = eval {
		 	module_dir($class);
		};
		next if $@;
		my $path = File::Spec->catfile($dir, $file);
		unless ( -e $path ) {
			next;
		}
		unless ( -r $path ) {
			Carp::croak("File '$file' cannot be read, no read permissions");
		}
		return $path;
	}
	Carp::croak("File '$file' does not exist in class or parent shared files");
}

#####################################################################
# Support Functions

sub _dist_packfile {
	my $module = shift;
	my @dirs   = grep { -e } ( $Config::Config{archlibexp}, $Config::Config{sitearchexp} );
	my $file   = File::Spec->catfile(
		'auto', split( /::/, $module), '.packlist',
	);

	foreach my $dir ( @dirs ) {
		my $path = File::Spec->catfile( $dir, $file );
		next unless -f $path;

		# Load the file
		my $packlist = ExtUtils::Packlist->new($path);
		unless ( $packlist ) {
			die "Failed to load .packlist file for $module";
		}

		die "CODE INCOMPLETE";
	}

	die "CODE INCOMPLETE";
}

# Inlined from Params::Util pure perl version
sub _CLASS {
    (defined $_[0] and ! ref $_[0] and $_[0] =~ m/^[^\W\d]\w*(?:::\w+)*\z/s) ? $_[0] : undef;
}


# Maintainer note: The following private functions are used by
#                  File::ShareDir::PAR. (It has to or else it would have to copy&fork)
#                  So if you significantly change or even remove them, please
#                  notify the File::ShareDir::PAR maintainer(s). Thank you!    

# Matches a valid distribution name
### This is a total guess at this point
sub _DIST {
	if ( defined $_[0] and ! ref $_[0] and $_[0] =~ /^[a-z0-9+_-]+$/is ) {
		return shift;
	}
	Carp::croak("Not a valid distribution name");
}

# A valid and loaded module name
sub _MODULE {
	my $module = _CLASS(shift) or Carp::croak("Not a valid module name");
	if ( Class::Inspector->loaded($module) ) {
		return $module;
	}
	Carp::croak("Module '$module' is not loaded");
}

# A valid file name
sub _FILE {
	my $file = shift;
	unless ( defined $file and ! ref $file and length $file ) {
		Carp::croak("Did not pass a file name");
	}
	if ( File::Spec->file_name_is_absolute($file) ) {
		Carp::croak("Cannot use absolute file name '$file'");
	}
	$file;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

File::ShareDir2 - Locate per-dist and per-module shared files

=head1 VERSION

version 0.001000

=head1 SYNOPSIS

  use File::ShareDir2 ':ALL';
  
  # Where are distribution-level shared data files kept
  $dir = dist_dir('File-ShareDir2');
  
  # Where are module-level shared data files kept
  $dir = module_dir('File::ShareDir2');
  
  # Find a specific file in our dist/module shared dir
  $file = dist_file(  'File-ShareDir2',  'file/name.txt');
  $file = module_file('File::ShareDir2', 'file/name.txt');
  
  # Like module_file, but search up the inheritance tree
  $file = class_file( 'Foo::Bar', 'file/name.txt' );

=head1 DESCRIPTION

The intent of L<File::ShareDir2> is to provide a companion to
L<Class::Inspector> and L<File::HomeDir>, modules that take a
process that is well-known by advanced Perl developers but gets a
little tricky, and make it more available to the larger Perl community.

Quite often you want or need your Perl module (CPAN or otherwise)
to have access to a large amount of read-only data that is stored
on the file-system at run-time.

On a linux-like system, this would be in a place such as /usr/share,
however Perl runs on a wide variety of different systems, and so
the use of any one location is unreliable.

Perl provides a little-known method for doing this, but almost
nobody is aware that it exists. As a result, module authors often
go through some very strange ways to make the data available to
their code.

The most common of these is to dump the data out to an enormous
Perl data structure and save it into the module itself. The
result are enormous multi-megabyte .pm files that chew up a
lot of memory needlessly.

Another method is to put the data "file" after the __DATA__ compiler
tag and limit yourself to access as a filehandle.

The problem to solve is really quite simple.

  1. Write the data files to the system at install time.
  
  2. Know where you put them at run-time.

Perl's install system creates an "auto" directory for both
every distribution and for every module file.

These are used by a couple of different auto-loading systems
to store code fragments generated at install time, and various
other modules written by the Perl "ancient masters".

But the same mechanism is available to any dist or module to
store any sort of data.

=head2 Using Data in your Module

C<File::ShareDir2> forms one half of a two part solution.

Once the files have been installed to the correct directory,
you can use C<File::ShareDir2> to find your files again after
the installation.

For the installation half of the solution, see L<Module::Install>
and its C<install_share> directive.

=head1 FUNCTIONS

C<File::ShareDir2> provides four functions for locating files and
directories.

For greater maintainability, none of these are exported by default
and you are expected to name the ones you want at use-time, or provide
the C<':ALL'> tag. All of the following are equivalent.

  # Load but don't import, and then call directly
  use File::ShareDir2;
  $dir = File::ShareDir2::dist_dir('My-Dist');
  
  # Import a single function
  use File::ShareDir2 'dist_dir';
  dist_dir('My-Dist');
  
  # Import all the functions
  use File::ShareDir2 ':ALL';
  dist_dir('My-Dist');

All of the functions will check for you that the dir/file actually
exists, and that you have read permissions, or they will throw an
exception.

=head2 dist_dir

  # Get a distribution's shared files directory
  my $dir = dist_dir('My-Distribution');

The C<dist_dir> function takes a single parameter of the name of an
installed (CPAN or otherwise) distribution, and locates the shared
data directory created at install time for it.

Returns the directory path as a string, or dies if it cannot be
located or is not readable.

=head2 module_dir

  # Get a module's shared files directory
  my $dir = module_dir('My::Module');

The C<module_dir> function takes a single parameter of the name of an
installed (CPAN or otherwise) module, and locates the shared data
directory created at install time for it.

In order to find the directory, the module B<must> be loaded when
calling this function.

Returns the directory path as a string, or dies if it cannot be
located or is not readable.

=head2 dist_file

  # Find a file in our distribution shared dir
  my $dir = dist_file('My-Distribution', 'file/name.txt');

The C<dist_file> function takes two params of the distribution name
and file name, locates the dist dir, and then finds the file within
it, verifying that the file actually exists, and that it is readable.

The filename should be a relative path in the format of your local
filesystem. It will simply added to the directory using L<File::Spec>'s
C<catfile> method.

Returns the file path as a string, or dies if the file or the dist's
directory cannot be located, or the file is not readable.

=head2 module_file

  # Find a file in our module shared dir
  my $dir = module_file('My::Module', 'file/name.txt');

The C<module_file> function takes two params of the module name
and file name. It locates the module dir, and then finds the file within
it, verifying that the file actually exists, and that it is readable.

In order to find the directory, the module B<must> be loaded when
calling this function.

The filename should be a relative path in the format of your local
filesystem. It will simply added to the directory using L<File::Spec>'s
C<catfile> method.

Returns the file path as a string, or dies if the file or the dist's
directory cannot be located, or the file is not readable.

=head2 class_file

  # Find a file in our module shared dir, or in our parent class
  my $dir = class_file('My::Module', 'file/name.txt');

The C<module_file> function takes two params of the module name
and file name. It locates the module dir, and then finds the file within
it, verifying that the file actually exists, and that it is readable.

In order to find the directory, the module B<must> be loaded when
calling this function.

The filename should be a relative path in the format of your local
filesystem. It will simply added to the directory using L<File::Spec>'s
C<catfile> method.

If the file is NOT found for that module, C<class_file> will scan up
the module's @ISA tree, looking for the file in all of the parent
classes.

This allows you to, in effect, "subclass" shared files.

Returns the file path as a string, or dies if the file or the dist's
directory cannot be located, or the file is not readable.

=head1 SEE ALSO

L<File::ShareDir::Install>, L<File::HomeDir>,
L<Module::Install>, L<Module::Install::Share>,
L<File::ShareDir::PAR>, L<Dist::Zilla::Plugin::ShareDir>

=head1 AUTHOR

Adam Kennedy <adamk@cpan.org>

=head1 CONTRIBUTORS

=for stopwords Graham Knop Hans Dieter Pearcey Jens Rehsack Kenichi Ishigaki Kent Fredric Ricardo SIGNES Steffen Müller Tatsuhiko Miyagawa

=over 4

=item *

Graham Knop <haarg@haarg.org>

=item *

Hans Dieter Pearcey <hdp@cpan.org>

=item *

Jens Rehsack <sno@netbsd.org>

=item *

Kenichi Ishigaki <ishigaki@cpan.org>

=item *

Kent Fredric <kentnl@cpan.org>

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Steffen Müller <smueller@cpan.org>

=item *

Tatsuhiko Miyagawa <miyagawa@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by Adam Kennedy <adamk@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
	qw(
		Laver Anomalio CSS
		)
);
for my $var (@stopwords) {
	$bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $var ) );
}
$bundle->add_or_append_policy_field(
	'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
$bundle->remove_policy('TestingAndDebugging::ProhibitNoStrict');
$bundle->remove_policy('RegularExpressions::RequireExtendedFormatting');
$bundle->remove_policy('RegularExpressions::RequireDotMatchAnything');
$bundle->remove_policy('Modules::ProhibitAutomaticExportation');
$bundle->remove_policy('Bangs::ProhibitDebuggingModules');

#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
	my $rcfile = path('./perlcritic.rc')->openw_utf8;
	$rcfile->print( $config->as_ini, "\n" );
	close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
	my $target = path('./misc');
	$target->mkpath if not $target->is_dir;

	my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
	for my $key ( sort keys %{$deps} ) {
		$depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
		*STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
	}
	close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/File/ShareDir2/Resolver.pm
use 5.006;    # our
use strict;
use warnings;

package File::ShareDir2::Resolver;

# ABSTRACT: A base class for Shared Path resolvers

our $VERSION = '0.001000';

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

sub new {
	my ( $package, @args ) = @_;
	use warnings FATAL => 'misc';    # Error on unbalanced @args
	return bless { ref $args[0] ? %{ $args[0] } : @args }, $package;
}

sub dist_dir {
  my ( $self, $distname ) = @_;
  return;
}

sub dist_file {
	my ( $self, $distname, $filename ) = @_;
	my $distdir = $self->dist_dir($distname);
	return unless defined $distdir;
	require File::Spec;
	my $path = File::Spec->catfile( $distdir, $filename );
	return unless -e $path;
	unless ( -f $path ) {
		Carp::croak("Found dist_file '$path', but not a file");
	}
	unless ( -r $path ) {
		Carp::croak("File '$path', no read permissions");
	}
	return $path;
}

sub module_dir {
  my ( $self, $module_name ) = @_;
  return;
}

sub module_file {
	my ( $self, $modulename, $filename ) = @_;
	my $moduledir = $self->module_dir($modulename);
	return unless defined $moduledir;
	require File::Spec;
	my $path = File::Spec->catfile( $moduledir, $filename );
	return unless -e $path;
	unless ( -r $path ) {
		Carp::croak("File '$filename' cannot be read, no read permissions");
	}
	return $path;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

File::ShareDir2::Resolver - A base class for Shared Path resolvers

=head1 VERSION

version 0.001000

=head1 AUTHOR

Adam Kennedy <adamk@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by Adam Kennedy <adamk@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/File/ShareDir2/Resolver/Perl.pm
use 5.006;
use strict;
use warnings;

package File::ShareDir2::Resolver::Perl;

# ABSTRACT: ShareDir 1.0 @INC resolver

our $VERSION = '0.001000';

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Carp qw( croak );
require File::ShareDir2::Resolver;
our @ISA = ('File::ShareDir2::Resolver');

sub dist_dir {
	my ( $self, $distname ) = @_;
	require File::Spec;
	my $path = File::Spec->catdir( 'auto', 'share', 'dist', $distname );

	# Find the full dir withing @INC
	foreach my $inc (@INC) {
		next unless defined $inc and !ref $inc;
		my $dir = File::Spec->catdir( $inc, $path );
		unless ( -d $dir ) {
			next;
		}

		unless ( -r $dir ) {
			croak("Found directory '$dir', but no read permissions");
		}
		return $dir;
	}
	return;
}

sub module_dir {
	my ( $self, $module ) = @_;
	require File::Spec;
	my $module_path = $module;
	$module_path =~ s/::/-/g;
	my $path = File::Spec->catdir( 'auto', 'share', 'module', $module_path );

	# Find the full dir withing @INC
	foreach my $inc (@INC) {
		next unless defined $inc and !ref $inc;
		my $dir = File::Spec->catdir( $inc, $path );
		unless ( -d $dir ) {
			next;
		}
		unless ( -r $dir ) {
			croak("Found directory '$dir', but no read permissions");
		}
		return $dir;
	}
	return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

File::ShareDir2::Resolver::Perl - ShareDir 1.0 @INC resolver

=head1 VERSION

version 0.001000

=head1 AUTHOR

Adam Kennedy <adamk@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by Adam Kennedy <adamk@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/File/ShareDir2/Resolver/Perl/Legacy.pm
use 5.006;    # our
use strict;
use warnings;

package File::ShareDir2::Resolver::Perl::Legacy;

# ABSTRACT: The Pre-1.0 resolver mechanic

our $VERSION = '0.001000';

# AUTHORITY

use Carp qw( croak );
require File::ShareDir2::Resolver;
our @ISA = ('File::ShareDir2::Resolver');
use constant IS_MACOS => !!( $^O eq 'MacOS' );

sub dist_dir {
	my ( $self, $distname ) = @_;

	# Create the subpath
	my $path = File::Spec->catdir( 'auto', split( /-/, $distname ), );

	# Find the full dir within @INC
	foreach my $inc (@INC) {
		next unless defined $inc and !ref $inc;
		my $dir = File::Spec->catdir( $inc, $path );
		next unless -d $dir;
		unless ( -r $dir ) {
			Carp::croak("Found directory '$dir', but no read permissions");
		}
		return $dir;
	}
	return;
}

sub dist_file {
	my ( $self, $distname, $filename ) = @_;

	# Create the subpath
	my $path = File::Spec->catfile( 'auto', split( /-/, $distname ), $filename, );

	# Find the full dir withing @INC
	foreach my $inc (@INC) {
		next unless defined $inc and !ref $inc;
		my $full = File::Spec->catdir( $inc, $path );
		next unless -e $full;
		unless ( -r $full ) {
			Carp::croak("Directory '$full', no read permissions");
		}
		return $full;
	}
	return;
}

sub module_dir {
	my ( $self, $module ) = @_;
	require Class::Inspector;
	my $short = Class::Inspector->filename($module);
	my $long  = Class::Inspector->loaded_filename($module);
	$short =~ tr{/}{:} if IS_MACOS;
	substr( $short, -3, 3, '' );
	$long =~ m/^(.*)\Q$short\E\.pm\z/s or die("Failed to find base dir");
	my $dir = File::Spec->catdir( "$1", 'auto', $short );

	unless ( -d $dir ) {
		Carp::croak("Directory '$dir', does not exist");
	}
	unless ( -r $dir ) {
		Carp::croak("Directory '$dir', no read permissions");
	}
	return $dir;
}

1;



Makefile.PL
# This Makefile.PL for File-ShareDir2 was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.34.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.006;
use ExtUtils::MakeMaker;

use lib 'inc/File-ShareDir-Install';

use File::ShareDir::Install;
install_share dist => "share";
install_share module => "File::ShareDir2", "foo";

my %WriteMakefileArgs = (
  "ABSTRACT" => "Locate per-dist and per-module shared files",
  "AUTHOR" => "Adam Kennedy <adamk\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "File-ShareDir2",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "File::ShareDir2",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Inspector" => 0,
    "Exporter" => 0,
    "File::Spec" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.89",
    "vars" => 0
  },
  "VERSION" => "0.001000",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);

my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::Inspector" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => "0.89",
  "constant" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}
